### ISSUE TITLE
Question - NAT traversal - is the secret sauce the low TTL in the packets when performing RENDEZVOUS? 

### ISSUE THREAD

**Issue by fanick1** [zerotier/ZeroTierOne#1264]
https://github.com/zerotier/ZeroTierOne/blob/0767ea5d693f6c79874e0ddfbc6b4796e64e6951/node/IncomingPacket.cpp#L611

Hi, I'm currently trying to sort of implement a DiY p2p discovery over UDP for one online multiplayer game, I know my code should work with the least restrictive NATs. On the 'central contact server' I remember peers' public addresses & ports and provide each peer with the other peers contact details when they indicate they want to be connected but for some reason my implementation does not work with my current ISP's NAT - my public ports seem to jump around at random, they don't stay the same for the same socket opened when switching from communicating with the central server and the peers - both peers end up connecting to wrong ports (they change from the ports registered at the central server).  Also fun fact - in my networking conditions the only situation when ports stay the same is when contacting well known STUN servers OR communicating with a host via port 3478, but this is almost useless to me, unless I'll have public game peer/server running on that port but hidden behind a firewall (it will be able to call back other peers then).
But after trying out ZeroTier and seeing it has no problem with punching through that bloody thing of NAT I came to the bright idea to poke at how you guys do it. From what I read here and there in essence you are not doing anything much different from what I do - exchange peers addresses and ports that they have connected with on your servers. But for some reason your approach works and my doesn't - so I wanted to ask some questions and I hope you could answer them or give any hints.

 - is the low TTL on the junk packet in `IncomingPacket.cpp:611` that important? Does it somehow magically make the NAT to do  something differently? (I guess not)
 - or is it in the ports that the ZeroTier opens on the client? (9993, etc) (I guess not, only 9993 is the default, other port numbers are derived)
 - I also have IPv6, could that be that thanks to that the ZeroTier utilizes also IPv6 for the traversal, the NAT then behaves differently?
What am I missing here?

**Comment by fanick1** [comment]
Also one of the peers in my test actually has public IP address, so his (firewalled) port 9993 probably became available from the outside at some point. I was not able to punch through with this setup in my implementation (I used different port, but I don't think that plays any role)

**Comment by laduke** [comment]
I don't know about those specifics, but check out [libzt](https://github.com/zerotier/libzt/)

**Comment by glimberg** [comment]
>  my public ports seem to jump around at random, they don't stay the same for the same socket opened when switching from communicating with the central server and the peers 

This sounds an awful lot like Symmetric NAT - the bane of any p2p communication protocol like ZeroTier. Basically, each remote IP/port combo is given a different port on the router to communicate via.  If possible in your router, you may be able to map ports directly to circumvent this.  The only real solution to Symmetric NAT is to not use Symmetric NAT, however.

> I also have IPv6, could that be that thanks to that the ZeroTier utilizes also IPv6 for the traversal, the NAT then behaves differently?

IPv6 doesn't need nat because there are so many more addresses available.

**Comment by fanick1** [comment]
I have cracked this nut literally five minutes ago - the trick in my case is to use some higher ports (e.g. 15000 or 25000) on the public central server.
Originally I was using ports in the range 5900-6000 ( no special reason for that ). I have hacked together a little UDP client-server where the server opens three ports (5900, 15900,  25900) and the client starts to machine-gun-connect to these ports from sequence of local ports (5900-6000).
The server just prints out the address and port of the incoming connection. From the output it was obvious that my NAT won't reuse ports used to connect to the 5900, but will reuse/share ports (mapping) used to connect to 15900 and 25900.

So I can just crank up the port numbers by few thousands and I should be fine.

Problem solved, case closed, thanks everybody for valuable inputs.
